design/baremetal-operator/configmap_for_host_inspection_data.md

```diff
@@ -0,0 +1,193 @@
+The spec of HostData will contain the same fields & hardware information as currently
+under `status.hardware` of BareMetalHost.
+
+### Pivoting
+
+There are three options one can choose to pivot, objects excluded from Cluster
+API chain, to a target cluster:
+
+1. [`clusterctl.cluster.x-k8s.io/move-hierarchy`](https://github.com/kubernetes-sigs/cluster-api/blob/0e6428151f2a388b98d1e220a698b85940a5f290/cmd/clusterctl/api/v1alpha3/labels.go#L38) label should be set
+    on BareMetalHost CRD before issuing *`clusterctl move`* so that the entire hierarchy
+    (i.e. including HostData, Secrets and ConfigMaps) is pivoted.
+
+2. `clusterctl.cluster.x-k8s.io=""` and `cluster.x-k8s.io/provider="metal3"`
+    labels on HostData CRD to extend the list of provider objects.
+
+3. `clusterctl.cluster.x-k8s.io/move` label on HostData CRD. This will
+    not pivot child objects as in option 1.
+
+### Implementation Details/Notes/Constraints
+
+The life of HostData is dependent on BareMetalHost. When BareMetalHost is entering
+inspection state, the HostData will be created. Failure in reaching inspection state
+or having `inspect.metal3.io: disabled` annotation on BareMetalHost will force the
+operator to not create the HostData. Deleting a BareMetalHost will result in deletion
+of a HostData respectively.
+
+HostData should be immutable and this will be controlled via [Validating Webhook](https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#validatingadmissionwebhook) and with RBAC permissions.
+In case, when [re-inspection](https://github.com/metal3-io/baremetal-operator/blob/master/docs/inspectAnnotation.md) of a `available` BareMetalHost is requested,
+new HostData will be re-created, because updating the old HostData will be
+blocked by the Validating Webhook.
+
+Deprovisioning of the BareMetalHost must not delete the HostData. It can
+be deleted either when BareMetalHost is being deleted or when re-inspection
+is requested. To ensure HostData availability while BareMetalHost is provisioned,
+a finalizer should be set on the HostData to prevent its accidental deletion.
+
+Only operator will be granted with create permissions for HostData and everyone
+else will be given no write permissions. If the BareMetalHost was inspected but
+HostData object doesn't exist, operator will first try to copy it from the
+BareMetalHost status (if available) and vice versa. If not available in neither
+of them, new inspection will be enforced.
+
+#### Keeping inspection data in sync
+
+As mentioned above, we will have two copies of inspection data available. One in the Spec
+of HostData and the other in the Status of BareMetalHost. One of them will get its inspection
+data copied from the other. Thus, inspection data can be copied from HostData (if available)
+to BareMetalHost, or the vice versa. If there is delta in the inspection data while
+copying, HostData should take precedence and get copied to the Status of BareMetalHost.
```

Comment:
I would really make a sync between them whenever the pause annotation is applied so that we dont have any mismatched status before and after pivot.

Reply:
@andfasano yes that's right. This double source of inspection data management will get much easier in the future when we drop the `status.hardware` from BMH.

Reply:
I'm wondering at this point if it could be better to remove the `vice versa` part, in order to also to prepare the ground for the future drop of `status.hardware` - I understand that this could mean a more complex implementation as of now.

Reply:
Sure, I'm more than happy to consider hostData as the main source of truth for the inspection data. As you mentioned, while `status.hardware` still exists, it can serve as a backup for the hostData only when hostData needs to be recreated. 

I remember @zaneb suggesting making it clear how we keep hostData and `status.hardware` of BMH in sync. So, I hope @zaneb would also agree on this?
To clarify the idea is:
- HostData is the main source of inspection data;
- Inspection data in the hostData can be populated/copied from BMH status only if hostData is missing
- Until we remove the `status.hardware` from BMH, inspection data can be populated/copied from the hostData to the status of BMH in case it is missing. This part is temporary, and will be removed during the API bump.

Reply:
As also mentioned in a later [comment](https://github.com/metal3-io/metal3-docs/pull/208#discussion_r773191847), a new problem to handle is how to manage the case where HD suddenly disappears (maybe due a deletion). In any case I would not use the BMH status to regenerate it, either because HD will be the single source of truth, either because it be removed in future